### PR TITLE
[SPARK-28074][SS] Log warn message on possible correctness issue for multiple stateful operations in single query

### PR DESCRIPTION
**Pull request by HeartSaVioR**
## What changes were proposed in this pull request?

Please refer [the link on dev. mailing list](https://lists.apache.org/thread.html/cc6489a19316e7382661d305fabd8c21915e5faf6a928b4869ac2b4a@%3Cdev.spark.apache.org%3E) to see rationalization of this patch.

This patch adds the functionality to detect the possible correct issue on multiple stateful operations in single streaming query and logs warning message to inform end users.

This patch also documents some notes to inform caveats when using multiple stateful operations in single query, and provide one known alternative.

## How was this patch tested?

Added new UTs in UnsupportedOperationsSuite to test various combination of stateful operators on streaming query.

### CODE REVIEW

File: docs/structured-streaming-programming-guide.md
Comment (dongjoon-hyun) [inline docs/structured-streaming-programming-guide.md]: Hi, @HeartSaVioR .
Could you add the screenshot to the PR description, please?

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [inline docs/structured-streaming-programming-guide.md]: _a_ global watermark? it also shouldn't be code-formatted
impact on query -> impact a query.
How would it impact the query?

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [inline docs/structured-streaming-programming-guide.md]: This is probably an important topic but I'm not clear what the guidance is here. If my answers are what then what do I do? what do you mean fail to answer the question? can this be more direct? what's an example of a problem, cause and solution?

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [reply docs/structured-streaming-programming-guide.md]: Meh, I don't know if it matters. We can evaluate the text and formatting pretty well here. That said there are too many bullet-pointed sentences here. Can't it just be prose?

**Review by srowen — CHANGES_REQUESTED** [review]

File: docs/structured-streaming-programming-guide.md
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md]: Honestly I agree with @srowen regarding screenshot, as it doesn't draw some UI like table. They're just bullet points with proper indentation, and if we add it to PR description I should update the screenshot whenever I reflect review comments, which sounds redundant.

Regarding too many bullet-pointed sentences, I agree there're some unnecessary bullet points. I'll concat. Any feedbacks on nuance, writing style, etc are welcome, as I'm not native speaker of English.

File: docs/structured-streaming-programming-guide.md
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md]: Will fix on above two lines of review comment.

> How would it impact the query?

The answer is in below line - `Fail to answer above questions might lead to incorrect outputs - e.g. intermediate outputs being discarded.` Not sure the flow looks natural, or you might want to revise the format/flow of content.

File: docs/structured-streaming-programming-guide.md
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md]: https://issues.apache.org/jira/browse/SPARK-28094 is the example on problem, as I tried to help resolving that issue (in user mailing list) and found the correctness issue.

> If my answers are what then what do I do?

The set of questions are asking theirselves as whether they're fully understanding the details of watermark, because it could bring "unexpected" discard on rows between stateful operators, say, correctness issue.

So it's like an exam to prevent end users to "just do it as Spark allows it" without fully understanding what they're doing. If they can't answer with their query, they should find alternative approach.

> what do you mean fail to answer the question?

Literally. If they are having any small unsure thing about answering these question they fail. I know it's too hard on end users, but well, better than realizing incorrect outputs in production env.

> can this be more direct? what's an example of a problem, cause and solution?

I'm happy to do it but not sure how much I need to explain with details. If we want to have detailed explanation like providing example, maybe better to have individual chapter. It might be  long enough to use as a blog post. This is an advanced topic - the book "Streaming Systems" assigns a chapter for watermark - so may not be feasible to add the details to the guide page, but due to the importance we should mention it. So that looks to be a dilemma.

To be honest, there's no "solution", only workaround, since it's due to lack of feature - watermark propagation and stateful operator-wise watermark. https://issues.apache.org/jira/browse/SPARK-26655 tries to address this, but not much efforts are here due to lack of interests.

**Comment by HeartSaVioR** [comment]
Actually clearer solution is just preventing any possible pairs and documenting alternative approach in guide doc, as I don't expect many end users could answer these questions. I'm just careful because then I'm proposing to drop supporting existing features (though I'm unsure they have been used).

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [reply docs/structured-streaming-programming-guide.md]: Yes, this might be too much for the docs here, but can we explain more clearly what type of incorrect usage may cause what problem? so that users know what to look for, in both their causes and symptoms. Otherwise it's just kind of saying "there are problems here if you're not careful".

Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md]: > what type of incorrect usage may cause what problem

Incorrect usages can't be collected out as some types - Spark allows arbitrary operations on state via map/flatMapGroupsWithState, so can't enumerate the overall usages and picks up incorrect usages. Users are open to encounter the situation whenever they're using multiple stateful operations in a query, according to their query.

One symptom would be incorrect outputs as SPARK-26655 represents, but Spark doesn't provide any information to indicate the issue so I had to analyze the query via answering these questions for one of query in SPARK-26655.

Maybe we shouldn't refer the case as "incorrect usage" even they've encountered, since the issue is coming from design and limitation on watermark in structured streaming. It's not their fault and they're not doing wrong: Spark just doesn't support it.

As I commented below, maybe disallowing the cases (with guiding alternative) until we make it correct would be better approach, but I'd defer to committers/PMCs on how to handle this.

**Comment by gaborgsomogyi** [comment]
Not yet considered the problem in depth but as a warm-up question: Is it possible to block the scenario(exception) or print warning in such cases?

**Comment by HeartSaVioR** [comment]
> Is it possible to block the scenario(exception) or print warning in such cases?

If you meant can we let Spark analyze query and check whether global watermark brings correctness issue for the query, then not sure we can. 

Instead, Spark could measure number of discarded rows in each stateful operator - that said, #21617 may be reopened - which helps letting end users indicate the issue from their query. 

I don't think it's sufficient though, as they would recognize while running query and the query would already push incorrect outputs to sink, and they have to stop the query and do non-trivial work (analyze, take workaround) to resolve the issue.

**Comment by HeartSaVioR** [comment]
Now I'm trying to narrow down the issue...

As we've figured out global watermark is the root cause (to discard intermediate rows or evict rows from state incorrectly), the condition which **might** bring correctness issue is, 1) having multiple stateful operators 2) more than one stateful operators have watermark to discard input rows/evict states.

Note that the condition is applied to "Append mode". For "Update mode" and "Complete mode", the behavior on multiple stateful operators are not defined properly - most of cases it needs retraction to correct the outputs for given key which is not supported, hence we could just define the condition as having multiple stateful operators.

Please also note that not all the cases would produce incorrect outputs. That makes hard to say what is the best approach from Spark side to avoid the correctness issue. Maybe we have some options here:

1) Define the case as "unsupported" and throw error on the query which meets the condition
1-A) Add config to unlock the case if end users are 100% sure their query is safe.
2) Allow the case, but log warning message to let end users notice that their query might not be safe.
3) Just document caveats and don't restrict or notice in runtime.

Would like to hear your opinions about these options.

**Comment by gaborgsomogyi** [comment]
I've invested time and now I understand the issue. I can confirm Spark discards data.

> Instead, Spark could measure number of discarded rows in each stateful operator - that said, #21617 may be reopened - which helps letting end users indicate the issue from their query.
I don't think it's sufficient though

I agree with this. If the user will see the problem only after the queries started is not sufficient.
Let me think a bit about your further proposals...

**Comment by gaborgsomogyi** [comment]
After consideration I would vote on point 2 because of the following:
* Doc is not always read carefully (thought it's a must to have)
* Point 1 is a little bit overkill since not all the use-case end-up in dropped data

File: docs/structured-streaming-programming-guide.md
Comment (gaborgsomogyi) [inline docs/structured-streaming-programming-guide.md]: Not sure this will be understood by users (I know the details since played a lot with it but also a bit struggling :)). I would focus on global watermark and maybe the simplest discard example + mentioning this may apply other situations.

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing @jose-torres 
Could you please evaluate the options I proposed above (below link) and guide preferred one?
https://github.com/apache/spark/pull/24890#issuecomment-504914060

This would also require either #24936 or #21617 (if we prefer #21617 I'll rework and submit again).

**Comment by HeartSaVioR** [comment]
Ping.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
I've revisited this and revised the patch. 

Now Spark will log warning message if the query contains the pattern of data loss on intermediate outputs. I've updated docs as well to remove the content which would be too hard to understand from end users.

I'm not sure the nuance is proper, so please review this again and correct the sentences. Thanks!

**Comment by HeartSaVioR** [comment]
retest this, please

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [inline docs/structured-streaming-programming-guide.md]: any of below -> any of the following
are possibly having issue -> can have this issue

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [inline docs/structured-streaming-programming-guide.md]: Maybe back-tick-quote the method names?
state function -> the state function

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [inline docs/structured-streaming-programming-guide.md]: that state -> that the state
is append mode -> uses Append mode

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala]: Because you stop when the first problem is detected, how about just a while loop, so you don't have to keep examining the whole plan? maybe it's OK as it only saves time in case of a problem, which is rare.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala]: foreach on the Option instead of match?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala]: See above suggestions about editing this text

**Review by srowen — CHANGES_REQUESTED** [review]

File: docs/structured-streaming-programming-guide.md
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md]: I think we can just remove "In some circumstance", as that's expected behavior as always.

* Streaming aggregation in Append mode: only evicted rows are emitted by nature, as it should wait for inputs until watermark passes by, to ensure there's no more rows to aggregate with such key.
* Outer Join (any mode): it basically emits matched rows, but it also emits evicted rows if the row haven't matched with other side of row. That's what outer join is, so it emits evicted rows conditionally, but mostly expected.
* FlatMapGroupsWithState in Append mode: it strongly depends on the implementation of state function, but if state function respects the semantic of Append mode, it could high likely emit evicted rows to ensure there's no further input rows to affect emitted rows.

In fact, emitting evicted rows are tied to the semantic of Append mode.

> how can users know if this affects them? 

We'll log warning message during checking unsupported operations. We can easily change the behavior to block the query as unsupported as well, so it's up to the community (mostly committers/PMCs) decision.

> is it what is described below?

Yes. And with SPARK-24634 we can measure the number of late rows per stateful operator in runtime. (For now there's no information on late rows. Spark discards them with no metrics.) First stateful operator could have late rows, but following state operators shouldn't.

File: docs/structured-streaming-programming-guide.md
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md]: Yes it describes potential future change. I left the content because operator-wise watermark is what Spark is left behind so it might be better to provide some promise, but it's not even ongoing effort (may require a new SPIP) so omitting it would be better to not provide wrong signal. I'll omit it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala]: It would require to rewrite the logic TreeNode provides. Maybe we can just call `throwError` and catch the exception and log warn message instead if failWhenDetected is true. It can achieve same result.

File: docs/structured-streaming-programming-guide.md
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md]: Replaced `In some circumstance` with `In Append mode` for clarification.

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [inline docs/structured-streaming-programming-guide.md]: state function -> the state function
of output -> of the output

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [inline docs/structured-streaming-programming-guide.md]: can be discarded implies that it's OK to discard them. Are you saying "may be discarded"?
Then I'd say "if a stateful operation emits rows older ... note that these rows may be discarded"

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [inline docs/structured-streaming-programming-guide.md]: of global -> of a global
could bring correctness -> can potentially cause a correctness..

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [inline docs/structured-streaming-programming-guide.md]: log warning -> log a warning
such pattern -> such a pattern

But above don't you say that Spark can't check this?

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [inline docs/structured-streaming-programming-guide.md]: You refer to the below operations twice here

File: docs/structured-streaming-programming-guide.md
Comment (srowen) [inline docs/structured-streaming-programming-guide.md]: state function -> the state function

**Review by srowen — CHANGES_REQUESTED** [review]

File: docs/structured-streaming-programming-guide.md
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md]: > can be discarded implies that it's OK to discard them. Are you saying "may be discarded"?

Ah thanks for correcting the details. If former interprets as that nuance, latter is correct.

File: docs/structured-streaming-programming-guide.md
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md]: Above sentence means "Spark can't check the state function is implemented via semantics of output mode" because we don't restrict anything on implementing state function (even FlatMapGroupsWithState knows about output mode but doesn't do anything with output mode), whereas this is checking logical plan to find the pattern which is possible - assuming the state function is implemented via semantics of output mode.

File: docs/structured-streaming-programming-guide.md
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md]: Ah OK I didn't mean "following" as same as "below" but it's pretty confusing. As we mention "after" here, "following" should be removed. Thanks!

**Review by srowen — COMMENTED** [review]
The text looks good. Anyone else want to evaluate the test/claim here?

**Review by gaborgsomogyi — APPROVED** [review]
Just revisited and LGTM.

**Comment by srowen** [comment]
Merged to master

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!

**Comment by HeartSaVioR** [comment]
FYI: #24936 is a kind of "follow-up" PR (though it could be used independently). Please take a look at the description of PR in #24936 - I've tested with query which has correctness issue, and explained how #24936 helps identifying the issue.

**Pull request closed without merging**
